site/content/contribute/more-info/server/schema-migration-guide.md

```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
 
 We are using {{< newtabref href="https://github.com/go-morph/morph" title="morph" >}} for the migration engine. The tool has a library and also a CLI. Mattermost server imports the library to have programmatic access to morph functions. A developer can use the morph CLI tool to test whether their migrations are working properly. Please follow instructions in the morph documentation to use the morph CLI tool.
 
-If your migration will take more time on a larger data set, check if it is possible to extract it as a separate SQL query for the customer to run off-hours on their DB. This is usually possible with new indexes/columns. As a result, when the application is upgraded, the migration is a no-op because those columns/indexes are already added.
+### Analysis
+
+A rough analysis of our past schema migrations shows the following (some very early migrations were skipped which would be considered as base Mattermost):
+
+```
+CREATE INDEX - 489
+ALTER TABLE - 195
+  ADD COLUMN - 113
+  ALTER COLUMN - 51
+  DROP COLUMN - 25
+  ADD CONSTRAINT - 6
+DROP INDEX - 124
+CREATE TABLE - 60
+UPDATE - 19
+DELETE - 2
+```
+
+![Migrations distribution](/contribute/more-info/server/migrations_pie.png)
+
+We will go through each of these migration types and discuss how we can make it non-blocking. This is a lengthy document, so for those wanting to directly look at the executive summary, we present it right now. And then expand on each section in detail later.
+
+### Conclusions
+
+**MySQL 5.7**
+
+| Operation      | Table rewrite | Concurrent DML allowed |
+| -----------    | ------------- | ---------------------- |
+| CREATE INDEX   | NO<sup>1</sup>| YES (with limitations)<sup>1</sup>|
+| DROP INDEX     | NO            | YES                    |
+| ADD COLUMN     | YES           | YES<sup>2</sup>                   |
+| ALTER COLUMN   | YES           | NO                     |
+| DROP COLUMN    | YES           | YES                    |
+| ADD FK CONSTRAINT| NO          | NO (avoidable)<sup>3</sup>        |
+| ADD UNIQUE CONSTRAINT | NO     | YES                    |
+
+Note:
+1. Fulltext and spatial indexes take locks. Adding a fulltext index can also rewrite the table if there is no user-defined FTS_DOC_ID column, or in other words if a fulltext index is being added for the first time.
+2. Concurrent DML is also not permitted while adding an auto-increment column. However we don’t use it and are unlikely to use it in future. So I have made the cell green for now.
+3. Setting `foreign_key_checks` to false avoids table locks. But it also defeats the purpose of a foreign key.
+
+**MySQL 8**
```

Comment:
On a editorial note, shall we use a single table for mysql? The only difference is at adding column we can add the exception for that in the footnote like others.

Reply:
Sure